src/flamenco/vm/fd_vm_private.h

```diff
@@ -184,16 +196,87 @@ fd_vm_mem_cfg( fd_vm_t * vm ) {
    hit loads and pretty good ILP.
 
    fd_vm_mem_haddr_fast is when the vaddr is for use when it is already
-   known that the vaddr region has a valid mapping. */
+   known that the vaddr region has a valid mapping.
+   
+   These assumptions don't hold if direct mapping is enabled since input
+   region lookups become O(log(n)). */
+
+
+/* fd_vm_get_input_mem_region_idx returns the index into the input memory
+   region array with the largest region offset that is <= the offset that
+   is passed in.  This function makes NO guarantees about the input being
+   a valid input region offset; the caller is responsible for safely handling
+   it. */
+static inline ulong
+fd_vm_get_input_mem_region_idx( fd_vm_t const * vm, ulong offset ) {
+  uint left  = 0U;
+  uint right = vm->input_mem_regions_cnt - 1U;
+  uint mid   = 0U;
+
+  while( left<right ) {
+    mid = (left+right) / 2U;
+    if( offset>=vm->input_mem_regions[ mid ].vaddr_offset+vm->input_mem_regions[ mid ].region_sz ) {
+      left = mid + 1U;
+    } else {
+      right = mid;
+    }
+  }
+  return left;
+}
 
-FD_FN_PURE static inline ulong
-fd_vm_mem_haddr( FD_FN_UNUSED fd_vm_t const *  vm,
-                 ulong                         vaddr,
-                 ulong                         sz,
-                 ulong const *                 vm_region_haddr, /* indexed [0,6) */
-                 uint  const *                 vm_region_sz,    /* indexed [0,6) */
-                 FD_FN_UNUSED uchar            write,           /* 1 if the access is a write, 0 if it is a read */
-                 ulong                         sentinel ) {
+/* fd_vm_find_input_mem_region returns the translated haddr for a given
+   offset into the input region.  If an offset/sz is invalid or if an 
+   illegal write is performed, the sentinel value is returned. */
+static inline ulong
+fd_vm_find_input_mem_region( fd_vm_t const * vm, 
+                             ulong           offset,
+                             ulong           sz, 
+                             uchar           write,
+                             ulong           sentinel,
+                             uchar *         is_multi_region ) {
+
+  /* Binary search to find the correct memory region.  If direct mapping is not
+     enabled, then there is only 1 memory region which spans the input region. */
+  ulong region_idx = fd_vm_get_input_mem_region_idx( vm, offset );
```

Comment:
Should we return the sentinel value here if the region is not found?

Reply:
It get's handled later in the function

Reply:
Do you mind adding a comment explaining how? The offset can potentially be any value as it's user-provided.

Reply:
I have a comment, but I'll also add a note at the top of the function
```
    if( FD_UNLIKELY( region_idx==vm->input_mem_regions_cnt ) ) {
      return sentinel; /* Access is too large */
    }
```